### ISSUE TITLE
lein-nvd always uses the latest version of nvd-clojure

### ISSUE THREAD

**Issue by nickmbailey** [rm-hull/lein-nvd#38]
We've been on an older version of lein-nvd for awhile now and last weeks release of 1.1.0 suddenly broke our existing builds which was surprising. I did some digging and it looks like no matter what version of the 'lein-nvd' plugin you use, it grabs the latest version of 'nvd-clojure' which is what controls the 'dependency-check' version. This is easy to reproduce:

```console
[Nicks-MacBook-Pro:20:46:39 ] git$ lein new nvd-test
Generating a project called nvd-test based on the 'default' template.
The default template is intended for library projects, not applications.
To see other templates (app, plugin, etc), try `lein help new`.
[Nicks-MacBook-Pro:20:46:46 ] git$ cd nvd-test/
[Nicks-MacBook-Pro:20:46:51 ] nvd-test$ vim project.clj
[Nicks-MacBook-Pro:20:47:25 ] nvd-test$ cat project.clj
(defproject nvd-test "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.9.0"]]
  :plugins [[lein-nvd "0.5.2"]]
  :repl-options {:init-ns nvd-test.core})
[Nicks-MacBook-Pro:20:47:46 ] nvd-test$ lein nvd update
Retrieving lein-nvd/lein-nvd/0.5.2/lein-nvd-0.5.2.pom from clojars
Retrieving lein-nvd/lein-nvd/0.5.2/lein-nvd-0.5.2.jar from clojars
Retrieving nvd-clojure/nvd-clojure/1.1.1/nvd-clojure-1.1.1.pom from clojars
Retrieving org/owasp/dependency-check-core/5.0.0/dependency-check-core-5.0.0.pom from central
Retrieving org/owasp/dependency-check-parent/5.0.0/dependency-check-parent-5.0.0.pom from central
Retrieving org/owasp/dependency-check-utils/5.0.0/dependency-check-utils-5.0.0.pom from central
Retrieving nvd-clojure/nvd-clojure/1.1.1/nvd-clojure-1.1.1.jar from clojars
Retrieving org/slf4j/slf4j-api/1.7.26/slf4j-api-1.7.26.jar from central
Retrieving org/owasp/dependency-check-core/5.0.0/dependency-check-core-5.0.0.jar from central
Retrieving org/owasp/dependency-check-utils/5.0.0/dependency-check-utils-5.0.0.jar from central
Retrieving joda-time/joda-time/2.9.9/joda-time-2.9.9.jar from central
Retrieving org/slf4j/jcl-over-slf4j/1.7.15/jcl-over-slf4j-1.7.15.jar from central
[2019-06-17 20:48:19.599] INFO Engine - Checking for updates
...
```

First I cleared my local .m2 repo and then ran the above. You can see that it first downloads lein-nvd 0.5.2 which is what is specified for the plugin, but then downloads nvd-clojure 1.1.1 and hence dependency-check 5.0.0. The behavior is the same if you use a more recent version like 1.0.0 which should use dependency-check 4.0.2:

```console
[Nicks-MacBook-Pro:20:51:54 ] nvd-test$ cat project.clj
(defproject nvd-test "0.1.0-SNAPSHOT"
  :description "FIXME: write description"
  :url "http://example.com/FIXME"
  :license {:name "EPL-2.0 OR GPL-2.0-or-later WITH Classpath-exception-2.0"
            :url "https://www.eclipse.org/legal/epl-2.0/"}
  :dependencies [[org.clojure/clojure "1.9.0"]]
  :plugins [[lein-nvd "1.0.0"]]
  :repl-options {:init-ns nvd-test.core})
[Nicks-MacBook-Pro:20:52:00 ] nvd-test$ lein nvd update
Retrieving lein-nvd/lein-nvd/1.0.0/lein-nvd-1.0.0.pom from clojars
Retrieving lein-nvd/lein-nvd/1.0.0/lein-nvd-1.0.0.jar from clojars
Retrieving nvd-clojure/nvd-clojure/1.1.1/nvd-clojure-1.1.1.pom from clojars
Retrieving org/owasp/dependency-check-core/5.0.0/dependency-check-core-5.0.0.pom from central
Retrieving org/owasp/dependency-check-parent/5.0.0/dependency-check-parent-5.0.0.pom from central
Retrieving org/owasp/dependency-check-utils/5.0.0/dependency-check-utils-5.0.0.pom from central
Retrieving nvd-clojure/nvd-clojure/1.1.1/nvd-clojure-1.1.1.jar from clojars
Retrieving org/owasp/dependency-check-utils/5.0.0/dependency-check-utils-5.0.0.jar from central
Retrieving org/owasp/dependency-check-core/5.0.0/dependency-check-core-5.0.0.jar from central
...
```

I'm not really sure how lein gets the dependency on nvd-clojure since it isn't listed explicitly. I tested this behavior on both lein 2.8.3 and lein 2.9.1 (latest).

**Comment by nickmbailey** [comment]
Oh I see the issue.

https://github.com/rm-hull/lein-nvd/blob/1.0.0/plugin/src/leiningen/nvd.clj#L32

It's adding the nvd-clojure dep dynamically there and using the "RELEASE" special version string to always grab the latest. So is this the intended behavior? If it is I'd say it's definitely worth documenting in the README as it's very counter intuitive.

**Comment by nickmbailey** [comment]
Also, for what its worth, having the dependency on nvd-clojure set dynamically like that makes it pretty  hard to debug things. The nvd-clojure dep doesn't make it into the output of 'lein deps :plugin-tree' so its a bit hard to see where certain things are coming from. In my case, the new OSS Sonatype analyzer is throwing an exception during initialization due to an issue with the version of guava it is using. 

Also, just so these comments don't sound overly negative/critical, thanks a ton for putting lein-nvd out there! It's been a great tool for us.

**Comment by rm-hull** [comment]
> So is this the intended behavior?

Yes, this is intended behaviour, when I originally wrote this plugin, I heavily based it on cloverage (which I also worked on), but you are right, it should be documented behaviour, along the lines of https://github.com/cloverage/cloverage#lein

**Comment by bostonaholic** [comment]
This is also biting me in the ass for the past week.

**Comment by nickmbailey** [comment]
We downgraded the plugin to 1.0.0 and in our CI environment we run nvd like this:

```
NVD_VERSION=1.0.0 lein nvd check
```

Which pins the nvd-clojure version (and hence dependency-check version). In my head I've been pondering if this really needs to use leins 'eval-in-project' which is what causes the guava conflict between dependency-check and my project. Seems like it could evaluate outside of the project but still pass the project classpath to dependency check. That's probably worth a separate issue though.

**Comment by rm-hull** [comment]
Related: #10

**Comment by bostonaholic** [comment]
Is this now fixed via #10 ?

**Comment by rm-hull** [comment]
Yes, it is. Closing
